### PR TITLE
Add shared dropdown element

### PR DESCRIPTION
**Pull request by learykara**
Adds shared dropdown element to thinkful ui.

### CODE REVIEW

File: src/Dropdown.jsx [hunk 1]
```diff
@@ -0,0 +1,93 @@
+const React = require('react');
+const classnames = require('classnames');
+
+require('./dropdown.less');
+
+
+/**
+ * Shared dropdown menu element
+ * @property data {Array} of items to display in the dropdown,
```
Comment (ripleyaffect) [inline src/Dropdown.jsx]: I'd guess that each item needs something specific. Either break the item out into it's own component and document those props there or document the array objects here

File: src/Dropdown.jsx [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+ * @property defaultDisplay {String} of default value that should be displayed
+ * @property handleChange {Function} to handle dropdown click/change
+ */
+const Dropdown = React.createClass({
+
+  getInitialState() {
+    return {open: false};
+  },
+
+  getDefaultProps() {
```
Comment (ripleyaffect) [inline src/Dropdown.jsx]: `propTypes` is a great way to self-document and ensure the correct thing is being passed in as well

File: src/Dropdown.jsx [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    return {data: []};
+  },
+
+  componentDidMount() {
+    document.addEventListener('click', (e) => this._checkClickAway(e));
```
Comment (ripleyaffect) [inline src/Dropdown.jsx]: I think `global` is preferable? @evanrs?

File: src/Dropdown.jsx [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+  },
+
+  componentWillUnmount() {
+    document.removeEventListener('click', (e) => this._checkClickAway(e));
+  },
+
+  _checkClickAway(e) {
+    if (!e.target.hasAttribute('data-clickable')) {
```
Comment (ripleyaffect) [inline src/Dropdown.jsx]: note to go though the way material-ui handles this later, since this seems a little hacky

File: src/Dropdown.jsx [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+      this.setState({open: false});
+    }
+  },
+
+  _generateNodes() {
+    let {data} = this.props;
```
Comment (ripleyaffect) [inline src/Dropdown.jsx]: const

File: src/Dropdown.jsx [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    return data.map((item, ind) => {
+      return (
+        <p
+          className="dropdown-item"
+          id={ind}
+          data-slug={item.slug}>
```
Comment (ripleyaffect) [inline src/Dropdown.jsx]: camelCase

Comment (ripleyaffect) [inline src/Dropdown.jsx]: `data-slug` is too specific for this component following whatever convention is used in `select` elements (`value` I think)

File: src/Dropdown.jsx [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+  componentClickAway() {
+    this.setState({open: false});
+  },
+
+  render() {
+    const nodes = this._generateNodes();
+    let {selectedInd, data, defaultDisplay} = this.props;
+
+    let buttonClasses = classnames(
```
Comment (ripleyaffect) [inline src/Dropdown.jsx]: const. no need for linebreak here

File: src/Dropdown.jsx [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+      'button', 'button__white', 'dd-button');
+    let dropdownClasses = classnames(
```
Comment (ripleyaffect) [inline src/Dropdown.jsx]: const

File: src/Dropdown.jsx [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+      'dd-open',
+      {'hidden': !this.state.open});
+
+    return (
+      <button
+        className={buttonClasses}
```
Comment (ripleyaffect) [inline src/Dropdown.jsx]: second indent for visual separatio

File: src/Dropdown.jsx [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+        onClick={this._toggleOpen}
+        data-clickable>
+        {data[selectedInd] && data[selectedInd].displayName || defaultDisplay}
+        <div
+          className={dropdownClasses}
+          onClick={this._handleChange}>
+          {nodes}
```
Comment (ripleyaffect) [inline src/Dropdown.jsx]: you can just call `_generateNodes` here

File: webpack.hotload.js [hunk 11]
```diff
@@ -6,6 +6,7 @@ var webpack = require('webpack');
 var WebpackDevServer = require('webpack-dev-server');
 
 module.exports = function (webpackConfig, options) {
+    options = options || {};
```
Comment (ripleyaffect) [inline webpack.hotload.js]: :+1:

**Comment by ripleyaffect** [comment]
Looks pretty good to me. Let's go over the open questions together and with @evanrs so we can get this out :dancer:

File: src/Dropdown.jsx (hunk 1 above)
Comment (learykara) [inline src/Dropdown.jsx]: array objects documented on L10

File: src/Dropdown.jsx (hunk 2 above)
Comment (learykara) [inline src/Dropdown.jsx]: added

File: src/dropdown.less [hunk 12]
```diff
@@ -0,0 +1,56 @@
+@import '~tfstyleguide/vars';
+
+.button.button__white.dd-button {
+  width: 420px;
```
Comment (evanrs) [inline src/dropdown.less]: Seems like this should be a max-width of 320 for mobile. @tholex?

Comment (tholex) [inline src/dropdown.less]: max-width: 100%; so it shrinks when the container gets smaller than the explicit `width`?

Comment (learykara) [inline src/dropdown.less]: got it

File: src/Dropdown.jsx [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+ * @property defaultDisplay {String} of default value that should be displayed
+ * @property handleChange {Function} to handle dropdown click/change
+ */
+const Dropdown = React.createClass({
+
+  propTypes: {
+    data: React.PropTypes.array.isRequired,
+    selectedInd: React.PropTypes.number,
+    defaultDisplay: React.PropTypes.string,
+    handleChange: React.PropTypes.func.isRequired
```
Comment (evanrs) [inline src/Dropdown.jsx]: This should be `onChange`, in general we should try to mimic conventions we find in material-ui.

https://github.com/callemall/material-ui/blob/master/src/radio-button.jsx#L100-L101

File: src/Dropdown.jsx [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+      containing `value` and `displayName`
+ * @property selectedInd {Int} of the selected index in the list
+ * @property defaultDisplay {String} of default value that should be displayed
+ * @property handleChange {Function} to handle dropdown click/change
+ */
+const Dropdown = React.createClass({
+
+  propTypes: {
+    data: React.PropTypes.array.isRequired,
+    selectedInd: React.PropTypes.number,
```
Comment (evanrs) [inline src/Dropdown.jsx]: Why is this property name truncated?

File: src/Dropdown.jsx (hunk 3 above)
Comment (evanrs) [inline src/Dropdown.jsx]: I'll submit a PR for this to do something like this:
https://github.com/callemall/material-ui/blob/master/src/mixins/click-awayable.js

**Pull request merged**